### PR TITLE
Added RxGlobalLayoutWatcher and support v4 for sparsearraycompat dependency

### PR DESCRIPTION
**Pull request by gokhanbarisaker**
Hi, I wanted to add  a Rx version of ViewTreeObserver.OnGlobalLayoutListener.

Cheers,

### CODE REVIEW

**Comment by JakeWharton** [comment]
Thanks for sending the PR but I can't merge something like this at this time. For starters the code style and formatting is completely different from the rest of the project and it also lacks any kind of tests. In general, however, the behavior and API that's here is something which is very different from the rest of the library and is extremely generic and pluggable which is something that I do not want this early. The currently library is violently opinionated and I would like to keep it that way–even at the risk of alienating some users since the lack of doing so yields poor results (see: the current RxAndroid).

If you want you're free to tell me more about how you are using something like this so that I can understand the use case for the future, but for now I'm going to close this PR.

**Comment by gokhanbarisaker** [comment]
I created a project to demonstrate most basic use case. If you would like to check, you can find it [here](https://github.com/gokhanbarisaker/RxGlobalLayoutWatcherSample). Also here is an abstract from PhotoAdapter,

```java
    private static final String PATTERN = "http://placehold.it/{{w}}x{{h}}";
    private SparseArrayCompat<Subscription> imageLoadSubscriptions = new SparseArrayCompat<>();

    @Override
    public void onBindViewHolder(final PhotoAdapter.ViewHolder viewHolder, int i) {
        Observable observable = RxGlobalLayoutWatcher.instance().addGlobalLayoutWatch(viewHolder.imageView, new RxGlobalLayoutWatcher.Mission(RxGlobalLayoutWatcher.OBJECTIVE_NONZEROAREA)).flatMap(new Func1<View, Observable<Boolean>>() {
            @Override
            public Observable<Boolean> call(View view) {
                Rect frame = new Rect(0, 0, view.getMeasuredWidth(), view.getMeasuredHeight());
                String path = TemplateFormatter.getInstance().format(PATTERN, frame);
                RequestCreator rc = Application.picasso.load(path);
                return RxPicasso.getInstance().loadImage(rc, viewHolder.imageView);
            }
        });

        imageLoadSubscriptions.append(viewHolder.imageView.hashCode(), observable.subscribe());
    }

    @Override
    public void onViewRecycled(ViewHolder holder) {
        super.onViewRecycled(holder);

        Subscription subscription = imageLoadSubscriptions.get(holder.imageView.hashCode());

        if (subscription != null)
        {
            subscription.unsubscribe();
            imageLoadSubscriptions.remove(holder.imageView.hashCode());
        }

        Application.picasso.cancelRequest(holder.imageView);
    }
```

If you like the idea, I would be glad to contribute. Also, this time with unit tests and code style ^^.

**Comment by JakeWharton** [comment]
An interesting usage! Do you know about Picasso's `.fit()` API? It will wait for the `imageView` to be measured before requesting an image at the exact size required?

**Comment by gokhanbarisaker** [comment]
As far as I understand, Picasso's .fit() API is optimizing bitmap size for better client memory usage. Still, we can only take advantage of it once we have a valid url and an ImageView. Although we have an ImageView in the snippet above, we are unlikely to receive an image without formatting the given *moustached size templates*, **{{w}}** and **{{h}}**.

Also, my goal here in the above snippet is decreasing the transferred bytes by taking advantage of our CDN capability of serving images with given sizes. Only after we have a measured view, we can generate a valid url.



### TL;DR;

Let's continue with a fictional example, and let's say I am polishing an application with lots of images.
My goal as a responsible developer is to create an application with,

  - Optimized network usage
  - Optimized CDN consumption
  - Optimized CDN consumption
  
  
#### Optimized network usage

Alright, we realized our original images are just too big for mobile clients, especially the ones with low resolution displays. Most of our clients already complaining of our slow, greedy and crashing application. e.g.,

  - "Images are not loading!",
  - "You just consumed half of my data plan quota in a day"
  - "I am giving 1 star to be visible in the Google Play..."
  - etc...


Moreover, we are receiving crash reports everyday with the OutOfMemoryException is the leader by far.


To solve this issue in the new version of our api, my awesome server-side team decided to provide me a url template. A template that has placeholders for preferred image dimensions. With that, each client can consume just the exact amount of data to display on UI.

    https://cdn.example.com/path/to/image/{{w}}x{{h}}

After the new update on our application, app starting to get a 5 star on the market, crash reports no-longer showing OutOfMemoryException like a nightmare in the top 10, everything looks fine, developers celebrating huge success...
Until server-side team starts to complain about not-getting the old performance from the CDN.


#### Optimized CDN consumption

It turns out our *responsive mobile design* combined with *android device ecosystem complexity* made impossible to make use of CDN cache with nearly each request adding new variation of same image.
While CDN performance is suffering from cache miss swarm, our server-side team requested from us to normalize our image dimensions to specific sizes.

    
    [500x500, 1000x1000, 1500x1500, ..., (n * 500) x (n * 500)];


and it worked. CDN is working as awesome as before, developers are happy once again, ...
until we once again started to receive complaints on the google play about crashes and our crash reports are once again showing up OutOfMemoryException in the top 10.



#### Optimized memory usage

I believe Picasso's .fit() API comes to play in this scenario, as a savior.
